File: designLinkedList.cpp
class MyLinkedList {
    
private:
    
    struct ListNode{
        int val;
        ListNode *next;
    };
    
    ListNode *head = NULL;
    
    
public:
    MyLinkedList() 
    {
        
    }
    
    int get(int index) {
        
        if(index >= length())
            return -1;
        
        ListNode *temp = head;
        
        for(int i = 0; i < index; i++)
        {
            temp = temp -> next;
        }
        
        return temp -> val;
    }
    
    void addAtHead(int val) {
        
        ListNode *newNode = new ListNode;
        newNode -> val = val;
        
        if(head == NULL)
        {
            head = newNode;
            newNode -> next = NULL;
            return;
        }
        
        newNode -> next = head;
        head = newNode;
    }
    
    void addAtTail(int val) 
    {
        ListNode *temp = head;
        ListNode *newNode = new ListNode;
        newNode -> val = val;
        newNode -> next = NULL;
        
        if(head == NULL)
        {
            head = newNode;
            return;
        }
        
        while(temp -> next != NULL)
        {
            temp = temp -> next;
        }
        
        
        temp -> next = newNode;
    }
    
    void addAtIndex(int index, int val) {
        
        ListNode *temp = head;
        ListNode *newNode = new ListNode;
        newNode -> val = val;
        
        if(index == length())
        {
             addAtTail(val);
            return;
        }
           
        
        if(index == 0)
        {
            addAtHead(val);
            return;
        }
        
        if(index > length())
            return;
        
        for(int i = 0; i < index - 1; i++)
        {
                
            temp = temp -> next;
        }
            
        newNode -> next = temp -> next;
        temp -> next = newNode;
        
    }
    
    void deleteAtIndex(int index) {
        
        ListNode *temp = head;
        
        if(index >= length())
            return;
        
        if(index == 0)
        {
            head = head -> next;
            return;
        }
        
        for(int i = 0; i < index - 1; i++)
        {
            temp = temp -> next;
        }
        
        ListNode *temp1 = temp -> next;
        temp -> next = temp1 -> next;
        delete temp1;
    }
    
    int length()
    {
        int length = 0;
        ListNode *temp = head;
        
        while(temp != NULL)
        {
            length++;
            temp = temp -> next;
        }

        return length ;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */

File: progress_sheet/designLinkedList.py
class Node:
    def __init__(self, val = None, next = None):
        self.val = val
        self.next = next

class MyLinkedList:

    def __init__(self):
        self.dummy = Node()
        self.head = self.dummy.next
        self.length = 0    

    def get(self, index: int) -> int:
        curr = self.dummy.next

        if index >= self.length:
            return -1

        for i in range(index):
            curr = curr.next
        
        return curr.val

        
    def addAtHead(self, val: int) -> None:
        curr = self.dummy
        curr.next = Node(val, curr.next)
        self.head = self.dummy.next
        self.length += 1

    def addAtTail(self, val: int) -> None:
        curr = self.dummy

        while curr and curr.next:
            curr = curr.next
        
        curr.next = Node(val)
        self.length += 1

    def addAtIndex(self, index: int, val: int) -> None:
        curr = self.dummy.next
        prev = self.dummy

        for i in range(index):
           
            if curr == None:
                return

            prev = curr
            curr = curr.next
        
        prev.next = Node(val, curr)
        self.head = self.dummy.next
        self.length += 1

    def deleteAtIndex(self, index: int) -> None:

        prev = self.dummy
        curr = self.dummy.next

        if index >= self.length:
            return

        for i in range(index):
            prev = curr
            curr = curr.next
        
        prev.next = curr.next
        self.head = self.dummy.next
        self.length -= 1

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
